test/controllers/products_controller_test.rb

```diff
@@ -0,0 +1,243 @@
+      end
+    end
+
+    describe "edit" do
+      it "can get to the product edit page" do
+        get edit_product_path(@product.id)
+        must_respond_with :success
+      end
+
+      it "will return not_found if product id is invalid" do
+        get edit_product_path(-1)
+        must_respond_with :not_found
+      end
+
+      it "does not allow a merchant to edit another merchant's product" do
+        post logout_path
+        perform_login(merchants(:merch_two))
+        @merchant = merchants(:merch_two)
+        get edit_product_path(@product.id)
+
+        expect(flash[:danger]).must_equal "You must be selling the product to edit it"
+
+        must_respond_with :redirect
+        must_redirect_to product_path(@product.id)
+      end
+    end
+
+    describe "update" do
+
+      let (:new_product_hash) {
+        {
+            product: {
+                name: "testing name name",
+                price: 25.50,
+                description: "testing description",
+                inventory_stock: 5,
+                photo_url: "github.com",
+                merchant_id: @merchant.id
+            }
+        }
+      }
+
+      it "can update an existing work" do
+        # product = products(:product_two)
+
+        expect {
+          patch product_path(@product.id), params: new_product_hash
+        }.wont_differ "Product.count"
+
+        must_redirect_to product_path(@product.id)
+
+        product = Product.find_by(id: @product.id)
+        expect(product.name).must_equal new_product_hash[:product][:name]
+      end
+
+      it "would show not_found given an invalid id" do
+        expect {
+          patch product_path(-1), params: new_product_hash
+        }.wont_differ "Product.count"
+
+        must_respond_with :not_found
+      end
+
+      it " will not update if the params is invalid" do
+        new_product_hash[:product][:name] = nil
+
+        expect {
+          patch product_path(@product.id), params: new_product_hash
+        }.wont_differ "Product.count"
+
+        @product.reload
+        must_respond_with :bad_request
+        expect(@product.name).wont_be_nil
+
+      end
+      # TODO: do we need to test for price validity
+    end
+
+    describe "toggle product status" do
+      it "can change product from active to retire" do
+        product = products(:product_one)
+        expect(product.status).must_equal "active"
+
+        patch update_product_status_path(product.id)
+        found_product = Product.find_by(name: "tree")
+        expect(found_product.status).must_equal "retired"
+      end
+
+      it "can change product status from retire to active" do
+        product = products(:product_seven)
+        expect(product.status).must_equal "retired"
+
+        patch update_product_status_path(product.id)
+        found_product = Product.find_by(name: "hot chocolate")
+        expect(found_product.status).must_equal "active"
+
+      end
+    end
+
+    # describe "destroy" do
```

Comment:
These are good tests, and they work if you put the `destroy` method back in. 

However you should include tests for:

- You have to be logged in and the owner of the product to destroy it.
- If you're not logged in you can't delete a product
- If you're logged in and not the owner you can't delete the product.